### PR TITLE
Feature/validation only node

### PR DESCRIPTION
**Pull request by liayoo**
DONE
- Introduced validation-only consensus nodes
  - Validators with producing right = whitelisted & staked validators that can become a block producer
  - Validators without producing right = non-whitelisted but staked validators that only has voting rights
- Introduced MAX_NUM_VALIDATORS and MAX_STAKE_PER_VALIDATOR
- Added consensus integration test

TODO
- Test with 15+ nodes!

### CODE REVIEW

File: consensus/index.js [hunk 1]
```diff
@@ -506,10 +507,10 @@ class Consensus {
     const prevBlock = number > 1 ? prevBlockInfo.block : prevBlockInfo;
 
     // Make sure we have at least MIN_NUM_VALIDATORS validators.
```
Comment (platfowner) [inline consensus/index.js:509]: Could you update this comment?

File: consensus/index.js [hunk 2]
```diff
@@ -977,20 +978,42 @@ class Consensus {
...
+      if (stake) {
+        if (whitelist[address] === true) {
+          if (MIN_STAKE_PER_VALIDATOR <= stake && stake <= MAX_STAKE_PER_VALIDATOR) {
+            validators[address] = { stake, producing_right: true };
+          }
+        } else {
+          candidates.push({
+            address,
+            stake,
+            expireAt: _.get(stakeInfo, `0.${PredefinedDbPaths.STAKING_EXPIRE_AT}`, 0)
```
Comment (platfowner) [inline consensus/index.js:998]: Just found that some labels (e.g. 'expire_at') are defined as constants but some others (e.g. 'stake', 'producing_right') are not. Do you have any criteria for that or just leaving them as todos?

File: consensus/index.js [hunk 3]
```diff
@@ -977,20 +978,42 @@ class Consensus {
...
+          });
+        }
       }
-    });
+    }
+    candidates = _.orderBy(candidates, ['stake', 'expireAt'], ['desc', 'desc']); // TODO(liayoo): How to do tie-breaking?
```
Comment (platfowner) [inline consensus/index.js:1003]: What about tie-breaking by addresses as they're random and unique as at least an interim solution?

File: blockchain/block.js [hunk 4]
```diff
@@ -134,6 +134,20 @@ class Block {
...
   }
 
+  static validateValidators(validators) {
+    if (!CommonUtil.isDict(validators)) return false;
+    for (const [address, info] of Object.entries(validators)) {
+      if (!CommonUtil.isCksumAddr(address)) {
+        return false;
+      }
+      if (!CommonUtil.isDict(info) || !CommonUtil.isNumber(info.stake) ||
+          !CommonUtil.isBool(info.producing_right)) {
```
Comment (platfowner) [inline blockchain/block.js:144]: just a suggestion: what about 'is_proposer' or 'can_propose' as they can propose blocks to produce blockchain with other validators?

File: genesis-configs/base/genesis_rules.json [hunk 5]
```diff
@@ -28,11 +28,11 @@
...
         "propose": {
-          ".write": "newData !== null && util.isDict(newData) && newData.proposer === auth.addr && Number($number) === newData.number && getValue('/consensus/whitelist/' + auth.addr) === true && (lastBlockNumber < 1 || getValue('/service_accounts/staking/consensus/' + auth.addr + '|0/balance') >= util.getMinStakeAmount()) && util.isNumber(newData.gas_cost_total) && (newData.gas_cost_total === 0 || newData.gas_cost_total === getValue('/service_accounts/gas_fee/gas_fee/' + $number + '/balance'))"
+          ".write": "util.isDict(newData) && newData.proposer === auth.addr && Number($number) === newData.number && getValue('/consensus/whitelist/' + auth.addr) === true && (lastBlockNumber < 1 || (getValue('/service_accounts/staking/consensus/' + auth.addr + '|0/balance') >= util.getMinStakeAmount() && getValue('/service_accounts/staking/consensus/' + auth.addr + '|0/balance') <= util.getMaxStakeAmount())) && util.isNumber(newData.gas_cost_total) && (newData.gas_cost_total === 0 || newData.gas_cost_total === getValue('/service_accounts/gas_fee/gas_fee/' + $number + '/balance'))"
```
Comment (platfowner) [inline genesis-configs/base/genesis_rules.json:31]: Maybe some util functions can be added to rule-util.js for:
- `getValue('/service_accounts/staking/consensus/' + auth.addr + '|0/balance')`
- `getValue('/service_accounts/gas_fee/gas_fee/' + $number + '/balance`
?

File: blockchain/block.js (hunk 4 above)
Comment (platfowner) [reply blockchain/block.js:144]: or proposal_right as it seems to be a general term?

File: consensus/index.js (hunk 1 above)
Comment (liayoo) [reply consensus/index.js]: Done!

File: consensus/index.js (hunk 2 above)
Comment (liayoo) [reply consensus/index.js:1004]: Defined constants for the labels. Thanks!

File: consensus/index.js (hunk 3 above)
Comment (liayoo) [reply consensus/index.js]: Thanks for the suggestion!! Added the tie-breaker & a note.

File: blockchain/block.js (hunk 4 above)
Comment (liayoo) [reply blockchain/block.js]: Renamed it to proposal_right. Thanks! 👍👍

File: genesis-configs/base/genesis_rules.json (hunk 5 above)
Comment (liayoo) [reply genesis-configs/base/genesis_rules.json:31]: Added util functions & updated other path getters in rule-util.js as well. Thanks!

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: genesis-configs/base/genesis_rules.json (hunk 5 above)
Comment (platfowner) [reply genesis-configs/base/genesis_rules.json:31]: Looks much clear, thanks!

File: db/rule-util.js [hunk 6]
```diff
@@ -130,10 +130,11 @@ class RuleUtil {
     return parsed[1];
   }
 
-  getAppAdminPath(accountName) {
+  isAppAdmin(accountName, address, getValue) {
     const { PredefinedDbPaths } = require('../common/constants');
     const appName = this.getServiceNameFromServAcntName(accountName);
-    return `/${PredefinedDbPaths.MANAGE_APP}/${appName}/${PredefinedDbPaths.MANAGE_APP_CONFIG}/${PredefinedDbPaths.MANAGE_APP_CONFIG_ADMIN}`;
+    return getValue(`/${PredefinedDbPaths.MANAGE_APP}/${appName}/${PredefinedDbPaths.MANAGE_APP_CONFIG}/` +
```
Comment (platfowner) [inline db/rule-util.js:136]: minor: what about `return !!getValue(...);`

File: db/rule-util.js [hunk 7]
```diff
@@ -146,14 +147,26 @@ class RuleUtil {
     }
   }
 
-  getBillingUserPath(billingServAcntName, userAddr) {
+  getBalance(addrOrServAcnt, getValue) {
+    return getValue(this.getBalancePath(addrOrServAcnt));
```
Comment (platfowner) [inline db/rule-util.js:151]: minor: what about `return getValue(...) || 0;`?

File: db/rule-util.js [hunk 8]
```diff
@@ -146,14 +147,26 @@ class RuleUtil {
...
+  }
+
+  isBillingUser(billingServAcntName, userAddr, getValue) {
     const { PredefinedDbPaths } = require('../common/constants');
     const parsed = this.parseServAcntName(billingServAcntName);
     const appName = parsed[1];
     const billingId = parsed[2];
-    return `/${PredefinedDbPaths.MANAGE_APP}/${appName}/${PredefinedDbPaths.MANAGE_APP_CONFIG}/` +
+    return getValue(
```
Comment (platfowner) [inline db/rule-util.js:159]: minor: ditto?

File: db/rule-util.js [hunk 9]
```diff
@@ -146,14 +147,26 @@ class RuleUtil {
...
+        `/${PredefinedDbPaths.MANAGE_APP}/${appName}/${PredefinedDbPaths.MANAGE_APP_CONFIG}/` +
         `${PredefinedDbPaths.MANAGE_APP_CONFIG_BILLING}/${billingId}/` +
-        `${PredefinedDbPaths.MANAGE_APP_CONFIG_BILLING_USERS}/${userAddr}`;
+        `${PredefinedDbPaths.MANAGE_APP_CONFIG_BILLING_USERS}/${userAddr}`);
+  }
+
+  getConsensusStakeBalance(address, getValue) {
```
Comment (platfowner) [inline db/rule-util.js:165]: minor: ditto?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks for the quick progress.

**Comment by liayoo** [comment]
Thanks for the review! Now merging..

**Pull request merged**